src/hotspot/share/gc/shenandoah/shenandoahScanRemembered.inline.hpp

```diff
@@ -553,15 +547,23 @@ ShenandoahScanRemembered<RememberedSet>::verify_registration(HeapWord* address,
     // This is a mixed evacuation: rely on mark bits to identify which objects need to be properly registered
 
     // If the object reaching or spanning the end of this card's memory is marked, then last_offset for this card
-    // should represents this object.  Otherwise, last_offset is a don't care.
-    HeapWord* end_of_interest = base_addr + max_offset;
+    // should represent this object.  Otherwise, last_offset is a don't care.
     do {
-      HeapWord* obj_addr = base_addr + offset;
-      oop obj = oop(base_addr + offset);
       prev_offset = offset;
-      offset = ctx->get_next_marked_addr(base_addr + offset, end_of_interest) - base_addr;
+      oop obj = cast_to_oop(base_addr + offset);
+      ShenandoahHeapRegion* region = heap->heap_region_containing(obj);
+      HeapWord* tams = ctx->top_at_mark_start(region);
```

Comment:
Think you can move both these lines outside the loop, because we're not crossing region boundaries here.

Reply:
Okay, wasn't sure if cards would always be aligned with regions.